pkg/isolationForest.py

```diff
@@ -0,0 +1,177 @@
+            }
+            entry["is_anomalypoint"] = False
+
+            row = anomalies_df[anomalies_df["uuid"] == uuid]
+            for metric in metric_columns:
+                if f"{metric}_pct_change" in anomalies_df.columns:
+                    entry["metrics"][metric]["percentage_change"] = float(
+                        row[f"{metric}_pct_change"].iloc[0]
+                    )
+                    if int(row["is_anomaly"].iloc[0]) == -1:
+                        entry["is_anomalypoint"] = True
+        return self.test["name"], dataframe_json
+
+    def output_text(self):
+        dataframe = self.dataframe
+        dataframe, anomalies_df = self.analyze(dataframe)
+        data_list = dataframe.values.tolist()
+        column_names = dataframe.columns.tolist()
+        tabulated_df = tabulate(data_list, headers=column_names)
+        formatted_table = self.format_dataframe(tabulated_df, anomalies_df, dataframe)
+        return self.test["name"], formatted_table
+
+    def analyze(self, dataframe: pd.DataFrame):
+        """Analyzing the data
+
+        Args:
+            dataframe (pd.DataFrame): _description_
+
+        Returns:
+            pd.Dataframe, pd.Dataframe: _description_
+        """
+        logger_instance = SingletonLogger(debug=logging.INFO).logger
+        logger_instance.info("Starting analysis using Isolation Forest")
+        metric_columns = [
+            column
+            for column in dataframe.columns
+            if column not in ["uuid", "buildUrl", "timestamp"]
+        ]
+        model = IsolationForest(contamination="auto", random_state=42)
+        dataframe_with_metrics = dataframe[metric_columns]
+        model = IsolationForest(contamination="auto", random_state=42)
+        model.fit(dataframe_with_metrics)
+        predictions = model.predict(dataframe_with_metrics)
+        dataframe["is_anomaly"] = predictions
+        anomaly_scores = model.decision_function(dataframe_with_metrics)
+
+        # Add anomaly scores to the DataFrame
+        dataframe["anomaly_score"] = anomaly_scores
+
+        # Calculate moving average for each metric
+        window_size = 5
+        moving_averages = dataframe_with_metrics.rolling(window=window_size).mean()
+
+        # Initialize percentage change columns for all metrics
+        for feature in dataframe_with_metrics.columns:
+            dataframe[f"{feature}_pct_change"] = 0.0
+
+        # Update DataFrame with percentage changes for anomalies exceeding 5%
+        for idx, row in dataframe.iterrows():
+            anomaly_check_flag = 0
+            if row["is_anomaly"] == -1:
+                for feature in dataframe_with_metrics.columns:
+                    pct_change = (
+                        (row[feature] - moving_averages.at[idx, feature])
+                        / moving_averages.at[idx, feature]
+                    ) * 100
+                    if abs(pct_change) > 5:
```

Comment:
I think we should allow this to be overwritten, and 5% change is quite noisy IMHO. I would argue for a 10-12% diff, wdyt?

Reply:
Let's have this too as a flag.. default to 10% should do the work. But still it's percentage change from the moving average so I'd be careful.